server/src/main/java/com/linecorp/centraldogma/server/command/AbstractCommandExecutor.java

```diff
@@ -125,6 +138,34 @@ public final void setWritable(boolean writable) {
         this.writable = writable;
     }
 
+    protected void throwExceptionIfRepositoryNotWritable(Command<?> command) throws Exception {
+        if (command instanceof NormalizableCommit) {
+            assert command instanceof RepositoryCommand;
+            final RepositoryCommand<?> repositoryCommand = (RepositoryCommand<?>) command;
+            if (InternalProjectInitializer.INTERNAL_PROJECT_DOGMA.equals(repositoryCommand.projectName())) {
+                return;
+            }
+            String repoName = repositoryCommand.repositoryName();
+            if (Project.REPO_META.equals(repoName)) {
+                // Use REPO_DOGMA for the meta repository because the meta repository will be removed.
+                repoName = Project.REPO_DOGMA;
```

Comment:
Question) Does it every make sense to mark `REPO_DOGMA` as not writable? If `REPO_DOGMA` becomes unwritable, how would one restore a repository to be writable? Would it make sense to just not allow making `REPO_DOGMA` unwritable instead?

Reply:
> how would one restore a repository to be writable? 

So, force push is used. 😉 
https://github.com/line/centraldogma/pull/1122/files#diff-ce9a71ae9a3b82069d8129da07529226ca4ced78c8ce6e6363f8f908e649edbcR129

Anyway, I realized I needed to make a few more changes. I’ll give you a ping once I’m done. 😉

Reply:
The reason I said this is due to how replay logs works:

1. Assume that a repository is already marksed as read only
2. We call `MetadataService#updateRepositoryStatus`
3. If there are any outstanding logs for `REPO_DOGMA`, an exception will always occur while trying to replay logs here: https://github.com/line/centraldogma/blob/f1ff2b65639cb47fa049a35d81ec1183eab8f91f/server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java#L1166
4. The `REPO_DOGMA` can't be updated via zookeeper logs - hence we are in a irrecoverable state.

Anyways, for most cases I guess this won't be an issue since I doubt we'll be marking `REPO_DOGMA` as read-only

Reply:
> an exception will always occur while trying to replay logs here:

We change to `PushAsIsCommand` before storing log:
https://github.com/line/centraldogma/blob/f1ff2b65639cb47fa049a35d81ec1183eab8f91f/server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java#L1176
Thus, when it's replaying `PushAsIsCommand`s are used.

StandAloneCommandExecutor raises an exception only when the command is instance of a normalize commit:
https://github.com/line/centraldogma/pull/1122/files#diff-1b79753a6234a28cbdd41ef181ff57c2c3ed6d439b4e7fb6092994f35359d7beR116

So exception won't be raised.

Reply:
What I wanted to say is that read-only mode prohibits a new commit but allows replayed commits. 😉

Reply:
> If there are any outstanding logs for REPO_DOGMA (we don't hold a log while checking whether a repo is read-only) , an exception will always occur while trying to replay logs here:

My point was rather that `throwExceptionIfRepositoryNotWritable` is not checked inside the distributed lock, so it's possible that a commit is made after the push from `updateRepositoryStatus` goes through.

The aforementioned commit will prevent additional commits from going through at `replayLogs`, hence an irrecoverable state.
This will only occur for `REPO_DOGMA`

Reply:
Nevermind this case since I doubt it will actually occur in production 🙇

Reply:
@minwoox explained offline.

Since the current implementation ignores `NormalizableCommit`s, commands executed by ZK won't be a target for `throwExceptionIfRepositoryNotWritable `.
Hence, 
1) commits that newly try to write to zk will fail due to `throwExceptionIfRepositoryNotWritable`
2) commits that have already been written to zk will always be processed since they aren't `NormalizableCommit`s

Reply:
Thanks for the clarification. 👍